tests/test_elements/test_mol2mol/test_mol2mol_filter.py

```diff
@@ -14,68 +22,172 @@
+        filtered_smiles_2 = pipeline.fit_transform(SMILES_LIST)
+        self.assertEqual(filtered_smiles_2, [SMILES_BENZENE, SMILES_CHLOROBENZENE])
+
+    def test_smarts_smiles_filter(self) -> None:
+        """Test if molecules are filtered correctly by allowed SMARTS patterns."""
+        smarts_pats: dict[str, Union[int, tuple[Optional[int], Optional[int]]]] = {
+            "c": (4, None),
+            "Cl": 1,
+        }
+        smarts_filter = SmartsFilter(smarts_pats)
+
+        smiles_pats: dict[str, Union[int, tuple[Optional[int], Optional[int]]]] = {
+            "c1ccccc1": (1, None),
+            "Cl": 1,
+        }
+        smiles_filter = SmilesFilter(smiles_pats)
+
+        for filter_ in [smarts_filter, smiles_filter]:
+            new_input_as_list = list(filter_.patterns.keys())
+            pipeline = Pipeline(
+                [
+                    ("Smiles2Mol", SmilesToMol()),
+                    ("SmartsFilter", filter_),
+                    ("Mol2Smiles", MolToSmiles()),
+                    ("ErrorFilter", ErrorFilter()),
+                ],
+            )
+            filtered_smiles = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(
+                filtered_smiles, [SMILES_BENZENE, SMILES_CHLOROBENZENE, SMILES_CL_BR]
+            )
+
+            pipeline.set_params(SmartsFilter__keep_matches=False)
+            filtered_smiles_2 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_2, [SMILES_ANTIMONY, SMILES_METAL_AU])
+
+            pipeline.set_params(
+                SmartsFilter__mode="all", SmartsFilter__keep_matches=True
+            )
+            filtered_smiles_3 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_3, [SMILES_CHLOROBENZENE])
+
+            pipeline.set_params(
+                SmartsFilter__keep_matches=True, SmartsFilter__patterns=["I"]
+            )
+            filtered_smiles_4 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_4, [])
+
+            pipeline.set_params(
+                SmartsFilter__keep_matches=False,
+                SmartsFilter__mode="any",
+                SmartsFilter__patterns=new_input_as_list,
+            )
+            filtered_smiles_5 = pipeline.fit_transform(SMILES_LIST)
+            self.assertEqual(filtered_smiles_5, [SMILES_ANTIMONY, SMILES_METAL_AU])
+
+    def test_smarts_filter_parallel(self) -> None:
+        """Test if molecules are filtered correctly by allowed SMARTS patterns in parallel."""
+        smarts_pats: dict[str, Union[int, tuple[Optional[int], Optional[int]]]] = {
+            "c": (4, None),
+            "Cl": 1,
+            "cc": (1, None),
+            "ccc": (1, None),
+            "cccc": (1, None),
+            "ccccc": (1, None),
+            "cccccc": (1, None),
+            "c1ccccc1": (1, None),
+            "cCl": 1,
+        }
+        smarts_filter = SmartsFilter(smarts_pats, mode="all", n_jobs=-1)
```

Comment:
We should keep the resource allocation to a minimum. So change `n_jobs=-1` to `n_jobs=2`.